src/utils/toposort.rs

```diff
@@ -1,25 +1,187 @@
-use std::collections::BTreeSet;
+#![allow(dead_code)] // Temporary until refactor is complete
 
-use portgraph::{algorithms as pg, Direction, NodeIndex, PortGraph, UnmanagedDenseMap};
+use std::hash::Hash;
 
-/// Topological sorting of the graph
+use itertools::Itertools;
+use petgraph::{visit::IntoNeighborsDirected, Direction};
+
+use crate::HashSet;
+
+/// Perform a topological sort of a graph, without reference to it.
+///
+/// This will only traverse the descendants of `root`.
```

Comment:
I think this next expanding. We require the graph to be a DAG and all nodes must be reachable from `root`.
(Otherwise you get annoying edge cases like `root` having predecessors).